homeassistant/components/panasonic_viera/config_flow.py

```diff
@@ -0,0 +1,165 @@
+                self._errors = {"base": ERROR_UNKNOWN}
+                return await self.async_step_user()
+
+            encryption_data = {
+                CONF_APP_ID: self._remote.app_id,
+                CONF_ENCRYPTION_KEY: self._remote.enc_key,
+            }
+
+            self._data = {**self._data, **encryption_data}
+
+            return self.async_create_entry(
+                title=self._data[CONF_NAME], data=self._data,
+            )
+
+        return self._show_pair_form()
+
+    async def async_step_import(self, import_config):
+        """Import a config entry from configuration.yaml."""
+        await self.async_load_data(import_config)
+        return await self.async_create_remote()
+
+    async def async_load_data(self, config):
+        """Load the data."""
+        self._data = config
+
+        self._data[CONF_PORT] = (
+            self._data[CONF_PORT] if CONF_PORT in self._data else DEFAULT_PORT
+        )
+        self._data[CONF_ON_ACTION] = (
+            self._data[CONF_ON_ACTION] if CONF_ON_ACTION in self._data else None
+        )
+
+        await self.async_set_unique_id(self._data[CONF_HOST])
+        self._abort_if_unique_id_configured()
+
+    async def async_create_remote(self):
+        """Create the remote."""
+        try:
+            self._remote = await self.hass.async_add_executor_job(
+                partial(RemoteControl, self._data[CONF_HOST], self._data[CONF_PORT])
+            )
+        except (TimeoutError, URLError, OSError) as err:
+            _LOGGER.error("Could not establish remote connection: %s", err)
+            self._errors = {"base": ERROR_NOT_CONNECTED}
+            return await self.async_step_user()
+        except Exception as err:  # pylint: disable=broad-except
+            _LOGGER.error("An unknown error occurred: %s", err)
+            self._errors = {"base": ERROR_UNKNOWN}
+            return await self.async_step_user()
+
+        if self._remote.type == TV_TYPE_ENCRYPTED:
+            return await self.async_step_pairing()
+
+        return self.async_create_entry(title=self._data[CONF_NAME], data=self._data,)
+
+    @callback
+    def _show_user_form(self):
+        """Show the initial form to the user."""
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_HOST,
+                        default=self._data[CONF_HOST]
+                        if self._data[CONF_HOST] is not None
+                        else "",
+                    ): str,
+                    vol.Optional(
+                        CONF_NAME,
+                        default=self._data[CONF_NAME]
+                        if self._data[CONF_NAME] is not None
+                        else DEFAULT_NAME,
+                    ): str,
+                }
+            ),
+            errors=self._errors,
+        )
+
+    @callback
+    def _show_pair_form(self):
```

Comment:
Maybe accept the errors as a parameter instead and just pass those along to show form? This method doesn't know who set the errors so it's weird that it should reset them.

Reply:
We normally don't store the errors as an instance attribute since they are specific to each step.

Reply:
yep, that makes sense

then can I set an `errors` argument (which defaults to False) in `async_step_user` and `async_step_pairing` too?

Reply:
No, our standard is to create an empty dict at the beginning of each step for `errors` variable. That's how it is reset.

Reply:
I now remember why I defined `_errors` as a class attribute

I wanted to make the remote restart the pairing process automatically if the PIN code was incorrect, so, instead of showing the form directly after an incorrect attempt, I would save the error to that attribute and call `async_step_user`, without the need for the user to press submit again.

Should I abandon this and restart the config flow entirely if the PIN code is incorrect?

Reply:
I could also just pass the errors through all of the functions, it could be a little messy but we would get rid of `_errors`

Reply:
Errors should we local to each step. If there's an error in a step, that step should show the form with the errors. We shouldn't route errors between steps.

Reply:
Consider dropping the form helper methods and just inline those in each step.

Reply:
I think we should inline the form helpers. Instead we can make a method that returns the data schema for the user form.